addons/hr_payroll/tests/test_benefit.py

```diff
@@ -0,0 +1,337 @@
+        })
+        self.richard_emp.resource_calendar_id = self.resource_calendar_id
+
+        self.benefit_type_leave = self.env['hr.benefit.type'].create({
+            'name': 'Leave',
+            'is_leave': True,
+            'code': 'LEAVE100'
+        })
+        self.leave_type = self.env['hr.leave.type'].create({
+            'name': 'Legal Leaves',
+            'time_type': 'leave',
+            'allocation_type': 'no',
+            'benefit_type_id': self.benefit_type_leave.id
+        })
+        self.benefit_type = self.env['hr.benefit.type'].create({
+            'name': 'attendance',
+            'is_leave': False,
+            'code': 'LEAVE100'
+        })
+
+    def test_no_duplicate(self):
+        self.richard_emp.generate_benefit(self.start, self.end)
+        pou1 = self.env['hr.benefit'].search_count([])
+        self.richard_emp.generate_benefit(self.start, self.end)
+        pou2 = self.env['hr.benefit'].search_count([])
+        self.assertEqual(pou1, pou2, "Benefits should not be duplicated")
+
+    def test_benefit(self):
+
+        self.richard_emp.generate_benefit(self.start, self.end)
+
+        attendance_nb = len(self.resource_calendar_id._attendance_intervals(self.start, self.end))
+        benefit_nb = self.env['hr.benefit'].search_count([('employee_id', '=', self.richard_emp.id)])
+        self.assertEqual(attendance_nb, benefit_nb, "One benefit should be generated for each calendar attendance")
+
+    def test_split_benefit_by_day(self):
+        start = datetime.strptime('2015-11-01 09:00:00', '%Y-%m-%d %H:%M:%S').replace(tzinfo=self.tz)
+        end = datetime.strptime('2015-11-03 18:00:00', '%Y-%m-%d %H:%M:%S').replace(tzinfo=self.tz)
+
+        # Benefit of type attendance should be split in three
+        benefit = self.env['hr.benefit'].create({
+            'name': '1',
+            'employee_id': self.richard_emp.id,
+            'date_start': start,
+            'date_stop': end,
+        })
+
+        benefits = benefit.split_by_day()
+        self.assertEqual(len(benefits), 3, "Benefit should be split in three")
+        self.assertEqual(pytz.utc.localize(benefits[0].date_start), self.tz.localize(datetime.strptime('2015-11-01 09:00:00', '%Y-%m-%d %H:%M:%S')))
+        self.assertEqual(pytz.utc.localize(benefits[0].date_stop), self.tz.localize(datetime.strptime('2015-11-01 23:59:59', '%Y-%m-%d %H:%M:%S')))
+
+        self.assertEqual(pytz.utc.localize(benefits[1].date_start), self.tz.localize(datetime.strptime('2015-11-02 00:00:00', '%Y-%m-%d %H:%M:%S')))
+        self.assertEqual(pytz.utc.localize(benefits[1].date_stop), self.tz.localize(datetime.strptime('2015-11-02 23:59:59', '%Y-%m-%d %H:%M:%S')))
+
+        self.assertEqual(pytz.utc.localize(benefits[2].date_start), self.tz.localize(datetime.strptime('2015-11-03 00:00:00', '%Y-%m-%d %H:%M:%S')))
+        self.assertEqual(pytz.utc.localize(benefits[2].date_stop), self.tz.localize(datetime.strptime('2015-11-03 18:00:00', '%Y-%m-%d %H:%M:%S')))
```

Comment:
please make helper functions for all that redundant localize stuff, it's ugly and unreadable